blockprotocol/blockprotocol
# Introduce improved Type System codegen utility for blocks

## 🌟 What is the purpose of this PR?

This introduces a completely rewritten codegen script for use in blocks. The new script is faster, more extensible, supports many more edge-cases, and is more customisable. 

Some of the notable improvements:

- Hoisting shared identifiers to a `shared.ts` file, this avoids many duplicated definitions of common types (especially data types such as `Text`)
- Removing duplicating documentation - the previous codegen utility could result in orphaned duplicated documentation strings lying around after their types (which were duplicates of others) were deleted
- The output is now idempotent, which makes the generation suitable for checking as part of CI (ensuring that the generated types stay up to date if the codegen utility is improved, or a user updates the type they're generating)
- Much better name conflict resolution. Previously we would identify duplicate entity type titles and append an incrementing number to them (or add a version to one of them), this instead intelligently differentiates conflicting type names by their class of type (data type, property type, entity type), their base URL, and their versions.
- Support for generating multiple types, to multiple files. Previously the codegen utility in `block-scripts` only supported generating a single schema for the source entity type.
- Link entities are now affirmed as having `linkData` (whereas previously this was left as optional)

This also fixes the dependency graph within the repo (at least related to `build` and `codegen`), which now allows us to use `codegen` rather than `schema` as the `yarn script` as the bug to do with site building has been fixed. This was due to us poorly saying that every single build required codegen, as opposed to individual ones.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1203179076056186/1203984686668764/f) _(internal)_
- [Asana task](https://app.asana.com/0/1203193968491715/1204155628030421/f) _(internal)_
- [Asana task](https://app.asana.com/0/1203179076056186/1204155628030424/f) _(internal)_
- [Asana task](https://app.asana.com/0/1203179076056186/1204155628030426/f) _(internal)_
- [Asana task](https://app.asana.com/0/1203179076056186/1203843256013361/f) _(internal)_

## 🚫 Blocked by

- [ ] Check spec for any updates that need to be made

## 🔍 What does this change?

Please see commits in order. It is a pretty huge PR but I was attempting to reach near-parity before submitting one. I have gone to considerable effort to make sure the commits are clean and help guide understanding in the design.

## 📜 Does this require a change to the docs?

- I'm unsure if we already mentioned the codegen in the docs, if we did we should update it
- As mentioned above, I think we do need to update the spec to mention the changes to the `package.json` as I believe we'd mentioned the `schema` keyword

## ⚠️ Known issues

- N/A

## 🐾 Next steps

- Generate a helper look-up map of URLs (versioned and perhaps base) to prettified constants so users can stop dealing with the long URLs and instead just use `name` or `columnId`, etc.
- Make the logging level configurable to blocks
- Add the type's IDs to the doc comments so that you know where the generated type came from
- Move the documentation from `<EntityName>Properties` to `<EntityName>` where it's more useful

## 🛡 What tests cover this?

- `tsc`
- `eslint`

These should be checked in the templates as part of CI now, whereas they weren't previously

## ❓ How to test this?

1.  Checkout the branch / view the deployment
1.  Try looking at the definition of the `CodegenParameters` type, and add some new elements to one of the block templates `package.json`
1. Look at the generated types
1. Also see [[DROP] - Example local usage](https://github.com/blockprotocol/blockprotocol/commit/72d7a2db00002f55f833d98e16fbaaff9ea56ed2) for an example (although this was done locally in `@blockprotocol/graph` as opposed to in one of the templates. It does show how we should update the `turbo.json` though)

## 📹 Demo

- [Use new codegen in React block template](https://github.com/blockprotocol/blockprotocol/commit/eda7aa5e8bb6df8d1c6d2cc993be51f8c6ed2fca)
- [Use new codegen in Custom element block template](https://github.com/blockprotocol/blockprotocol/commit/42d5d596c26ac9c7a6bfedd88100bff1c1958b9e)
- [[DROP] - Example local usage](https://github.com/blockprotocol/blockprotocol/commit/72d7a2db00002f55f833d98e16fbaaff9ea56ed2)
